session.go

```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
 				if !ok {
 					req.Reply(false, nil)
 					continue
 				}
-			}
-			sess.pty = &ptyReq
-			sess.winch = make(chan Window, 1)
-			sess.winch <- ptyReq.Window
-			defer func() {
-				// when reqs is closed
-				close(sess.winch)
-			}()
-			req.Reply(ok, nil)
-		case "x11-req":
-			if sess.handled || sess.x11 != nil {
-				req.Reply(false, nil)
-				continue
-			}
-			x11Req, ok := parseX11Request(req.Payload)
-			if !ok {
+				if sess.ptyCb != nil {
+					ok := sess.ptyCb(sess.ctx, ptyReq)
+					if !ok {
+						req.Reply(false, nil)
+						continue
+					}
+				}
+				sess.pty = &ptyReq
+				sess.winch = make(chan Window, 1)
+				sess.winch <- ptyReq.Window
+				defer func() {
+					// when reqs is closed
+					close(sess.winch)
+				}()
+				req.Reply(ok, nil)
+			case "x11-req":
+				if sess.handled || sess.x11 != nil {
+					req.Reply(false, nil)
+					continue
+				}
+				x11Req, ok := parseX11Request(req.Payload)
+				if !ok {
+					req.Reply(false, nil)
+					continue
+				}
+				sess.x11 = &x11Req
+				if sess.x11Cb != nil {
+					ok := sess.x11Cb(sess.ctx, x11Req)
+					req.Reply(ok, nil)
+					continue
+				}
 				req.Reply(false, nil)
-				continue
-			}
-			sess.x11 = &x11Req
-			if sess.x11Cb != nil {
-				ok := sess.x11Cb(sess.ctx, x11Req)
+			case "window-change":
+				if sess.pty == nil {
+					req.Reply(false, nil)
+					continue
+				}
+				win, _, ok := parseWindow(req.Payload)
+				if ok {
+					sess.pty.Window = win
+					sess.winch <- win
+				}
 				req.Reply(ok, nil)
-				continue
-			}
-			req.Reply(false, nil)
-		case "window-change":
-			if sess.pty == nil {
+			case agentRequestType:
+				// TODO: option/callback to allow agent forwarding
+				SetAgentRequested(sess.ctx)
+				req.Reply(true, nil)
+			case keepAliveRequestType:
+				if req.WantReply {
+					req.Reply(true, nil)
+				}
+			case "break":
+				ok := false
+				sess.Lock()
+				if sess.breakCh != nil {
+					sess.breakCh <- true
+					ok = true
+				}
+				req.Reply(ok, nil)
+				sess.Unlock()
+			default:
+				// TODO: debug log
 				req.Reply(false, nil)
-				continue
-			}
-			win, _, ok := parseWindow(req.Payload)
-			if ok {
-				sess.pty.Window = win
-				sess.winch <- win
 			}
-			req.Reply(ok, nil)
-		case agentRequestType:
-			// TODO: option/callback to allow agent forwarding
-			SetAgentRequested(sess.ctx)
-			req.Reply(true, nil)
-		case "break":
-			ok := false
-			sess.Lock()
-			if sess.breakCh != nil {
-				sess.breakCh <- true
-				ok = true
-			}
-			req.Reply(ok, nil)
-			sess.Unlock()
-		default:
-			// TODO: debug log
-			req.Reply(false, nil)
 		}
+
+	}
+}
+
+func KeepAliveRequestHandler(ctx Context, srv *Server, req *gossh.Request) (ok bool, payload []byte) {
+	log.Printf("Handle keep-alive request: %s (wantReply: %t)", req.Type, req.WantReply)
+
+	if !req.WantReply {
+		return true, nil
+	}
+
+	err := req.Reply(true, nil)
+	if err != nil {
+		log.Printf("Replying to client keep-alive request failed: %v", err)
```

Comment:
What verbosity will this show as? Do we need to worry about coderd logs filling with keepalive messages?

Reply:
> What verbosity will this show as?

Once we update it in Coder, I guess that it will go with the default log level. 

> Do we need to worry about coderd logs filling with keepalive messages?

Actually, it depends on the Coder agent configuration. In theory, we should set the keep-alive ping close to the agent timeout. If the agent timeout is 120m, then we can set the interval to 119m.